go/src/url_shortener/stats.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+// StatsJSON Statistic data ready for JSON serialization
+type StatsJSON struct {
+	ServerStats serverStatsJSON `json:"server_stats"`
+
+	mux sync.Mutex
+}
+
+type serverStatsJSON struct {
+	TotalURL  int           `json:"total_url"`
+	Redirects redirectsJSON `json:"redirects"`
+	Handlers  []handlerJSON `json:"handlers"`
+}
+
+type redirectsJSON struct {
+	Success int `json:"success"`
+	Failed  int `json:"failed"`
+}
+
+type handlerJSON struct {
+	Name  string `json:"name"`
+	Count int    `json:"count"`
+}
+
+func (s *StatsJSON) String() string {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	stats := s.ServerStats
+
+	pairsInCache := fmt.Sprintf("Number of long/short URL pairs: %v", stats.TotalURL)
+
+	redirects := stats.Redirects
+	succeededRedirects := fmt.Sprintf("Succeeded redirects: %v", redirects.Success)
+	failedRedirects := fmt.Sprintf("Failed redirects: %v", redirects.Failed)
+
+	handlers := &stats.Handlers
+	handlerCalls := make([]string, 0, len(*handlers))
+
+	for _, handler := range *handlers {
+		handlerCalls = append(handlerCalls, fmt.Sprintf("Handler %s called %v time(s)", handler.Name, handler.Count))
+	}
+
+	statsBody := fmt.Sprintf("Some statistics:\n\n%s\n%s\n%s\n%s", pairsInCache, succeededRedirects, failedRedirects, strings.Join(handlerCalls, "\n"))
+	return statsBody
```

Comment:
For this kind of case it's easier, but also more performance in terms of volume of allocations, to use a [strings.Builder](https://godoc.org/strings#Builder) example: https://play.golang.org/p/LnoC147qON5